### ISSUE TITLE
[Windows] Since 2.2b1 baksmali goes crazy with paths

### ISSUE THREAD

**Issue by mirh** [JesusFreke/smali#548]
\>java -jar baksmali-smali-2.2.1.jar x file.jar -o folder

````
Error occurred while loading class path files. Aborting.
org.jf.dexlib2.analysis.ClassPathResolver$ResolveException: org.jf.dexlib2.analysis.ClassPathResolver$NotFoundException: Could not find classpath entry /system/framework/core.jar
        at org.jf.dexlib2.analysis.ClassPathResolver.<init>(ClassPathResolver.java:136)
        at org.jf.dexlib2.analysis.ClassPathResolver.<init>(ClassPathResolver.java:106)
        at org.jf.baksmali.AnalysisArguments.loadClassPathForDexFile(AnalysisArguments.java:123)
        at org.jf.baksmali.AnalysisArguments.loadClassPathForDexFile(AnalysisArguments.java:89)
        at org.jf.baksmali.DisassembleCommand.getOptions(DisassembleCommand.java:199)
        at org.jf.baksmali.DeodexCommand.getOptions(DeodexCommand.java:71)
        at org.jf.baksmali.DisassembleCommand.run(DisassembleCommand.java:173)
        at org.jf.baksmali.Main.main(Main.java:102)
Caused by: org.jf.dexlib2.analysis.ClassPathResolver$NotFoundException: Could not find classpath entry /system/framework/core.jar
        at org.jf.dexlib2.analysis.ClassPathResolver.loadLocalOrDeviceBootClassPathEntry(ClassPathResolver.java:207)
        at org.jf.dexlib2.analysis.ClassPathResolver.<init>(ClassPathResolver.java:121)
        ... 7 more
````
Using ProcMon I noticed it tries to create file `C:\system\framework\core.jar` (full disclosure: I definitively wasn't using the tool from the disk root folder)
And file `%workingdir%\;system;framework;core.jar`, of course illegal

**Comment by JesusFreke** [comment]
it's trying to create the file? or read the file? The latter is actually WAI, although it certainly shouldn't be trying to create such a file.

Can you describe your environment a bit more? What is the current working directory when you execute that command, and where are the framework files that should be used for deodexing?

**Comment by mirh** [comment]
ProcMon says operation is "createfile". Don't remember what exactly this entails. 

Env is normal w7 64 bit. And folder is simple one on my desktop with file inside.

**Comment by JesusFreke** [comment]
https://msdn.microsoft.com/en-us/library/windows/desktop/aa363858(v=vs.85).aspx

    Creates or opens a file or I/O device.

So yeah, it's looking for the framework files that it needs to deodex the odex file. It tries to be smart and uses some heuristics to guess where they might be. But it sounds like you don't even have the framework files, so of course deodexing is going to fail.

I recommend pulling the framework folder from your device, and passing the path to that folder to baksmali via the -d option.

**Comment by mirh** [comment]
I have no problems with the file alone on 2.1.3

**Comment by JesusFreke** [comment]
Then it either didn't have any optimized instructions that needed to be deodexed (unlikely) or it wasn't being fully deodexed.

The framework files are required for deodexing.

**Comment by mirh** [comment]
As I said, they never were needed previously. 
Also, even then it's still not working.

**Comment by JesusFreke** [comment]
Please see "baksmali help classpath" to determine how to provide the framework classes that baksmali needs, which varies based on the version of Android. If you can't get it working, please show the exact command you are using, which android version the device is.

**Comment by mirh** [comment]
>java -jar baksmali-2.2.1.jar x [services.jar](https://github.com/JesusFreke/smali/files/1266509/services.zip) -d "whatever\android-sdk-windows\platform-tools\system\framework"

**Comment by JesusFreke** [comment]
Well, that's one piece of the requested information.

**Comment by mirh** [comment]
Is it so hard to download file and reproduce it yourself?
Android version is 4.1 and stack trace was already posted.

**Comment by JesusFreke** [comment]
The file you posted isn't even an odex file. It's a normal jar with a classes.dex file.

Even if it were an odex file, I still couldn't try it on my own, because I don't have the framework files from the device which you had pulled the odex from.

**Comment by JesusFreke** [comment]
Assuming that really is the file that you want to "deodex", it doesn't need to be deodexed, because it's not an odex file.

Just use "baksmali disassemble services.jar" instead.

**Comment by mirh** [comment]
You are right. 
Was it expected for x behavior to change?

**Comment by mirh** [comment]
Ie, the lack of [this check](https://github.com/JesusFreke/smali/blob/v2.1.3/baksmali/src/main/java/org/jf/baksmali/main.java#L304). 
I have seen the one on Disassemble command was kept in the rewriting.

**Comment by JesusFreke** [comment]
Requiring the classpath files when deodexing is WAI. If you want to explicitly not use a classpath, you can use --bootclasspath ""  (i.e. an empty boot classpath)

**Comment by mirh** [comment]
I was talking about the ignoring of deodex parameter, when un-needed. 
Which, I mean, was nonetheless dumb to use on my side .. But indeed, I was wondering if this behavior asymmetry with older versions was intended or not.

**Comment by JesusFreke** [comment]
Requiring the classpath files when deodexing is WAI... :)

**Comment by mirh** [comment]
Yes.. But I'm referring to the case where there is \*not\* something to dedox, like at all, *in the first place*

**Comment by JesusFreke** [comment]
Requiring the classpath files when using the deodex command (also known as... wait for it... "deodexing") is WAI.

**Comment by mirh** [comment]
But the check you once had in place in main (that to the best of my knowledge was executed unconditionally of everything else, ie before even attempting any deodexing) predates what you are talking about. 
I even linked the code!

**Comment by JesusFreke** [comment]
I'm not sure what part of "WAI" you're not getting here :)

**Comment by mirh** [comment]
I'm not getting the "classpath" thing actually. 
Being something deodex-related (right? - or at least in this case), it's placed *after* the program \*decides\* to branch into DeodexCommand. 
So, if my doubt is about something that should or should not happen early.. what's it?

**Comment by Lanchon** [comment]
lololol this is like my 27th notification email! on this :)

**Comment by JesusFreke** [comment]
tl;dr; while you can use a non-odex/non-oat file with the deodex command, there's no reason to do so, and I don't care about improving the user experience for that.

Deodexing involves translating optimized opcodes that an invalid in a normal dex file back to their normal unoptimized equivalents. These optimized opcodes themselves don't have all of the information needed to do the reverse transformation. So baksmali needs access to the full class hierarchy in order to reconstruct the information needed to convert these back to normal instructions.

For example, for the optimized invoke-virtual-quick instruction, the method being invoked is stored as a virtual-table index instead of a pointer to the method name and prototype like for invoke-virtual. So baksmali has to calculate the full vtable for the class, so it can find which method it is referring to. And to construct the vtable, baksmali needs access to the full class hierarchy. Which is why the deodex command requires the framework files from the device.

While it's possible to scan the input file to try and determine if it actually need the full classpath or not, there's no reason to use the deodex command with a file that doesn't need a classpath. So instead of taking the extra time needed to scan the dex file and determine if it needs the classpath, it just assumes it does.

**Comment by mirh** [comment]
Yes, I had kind of got what odex entailed. 
I was just wondering if _dropping_ hasOdexOpcodes check from DeodexCommand was WAI or not. 
Class paths answer seemed quite off-the-point.
